File: ios/VoteGlobal/Data/Blockchain/DIDService.swift
import Foundation
import CryptoKit

/**
 * DIDService — W3C Decentralised Identity (did:ethr) client for iOS.
 *
 * Generates a DID from the device's ephemeral session key and
 * derives the identity commitment for VoterRegistry registration.
 *
 * The private key never leaves the Secure Enclave.
 * The DID is never transmitted to the backend — only the commitment hash.
 */
final class DIDService {

    private let networkName: String
    private let keyStore: EncryptedKeyStore
    private let apiClient: VoteAPIClient

    init(networkName: String = "sepolia", keyStore: EncryptedKeyStore, apiClient: VoteAPIClient) {
        self.networkName = networkName
        self.keyStore = keyStore
        self.apiClient = apiClient
    }

    // MARK: - DID Generation

    /// Generate a did:ethr DID from the session public key (Ethereum address).
    func generateDID() throws -> String {
        guard let rawKey = keyStore.loadSessionKey() else {
            throw DIDError.noSessionKey
        }
        let privateKey = try P256.Signing.PrivateKey(rawRepresentation: rawKey)
        let address = ethereumAddress(from: privateKey.publicKey)
        return "did:ethr:\(networkName):\(address)"
    }

    // MARK: - Identity Commitment

    /**
     * Derive a voter identity commitment.
     * commitment = SHA-256(did || countryCode || salt)
     *
     * Only the commitment is sent to the server — the DID stays on-device.
     */
    func deriveIdentityCommitment(did: String, countryCode: String, salt: String) -> String {
        var hasher = SHA256()
        hasher.update(data: Data((did + countryCode + salt).utf8))
        return Data(hasher.finalize()).hexString
    }

    // MARK: - VC Verification (via backend oracle)

    /**
     * Register with the VoterRegistry using a government-issued
     * Verifiable Credential JWT.
     *
     * Flow:
     *   1. User receives a VC JWT from the national identity oracle
     *   2. App derives DID and salt locally
     *   3. App posts (vcJwt, countryCode, salt) to backend /did/commitment
     *   4. Backend verifies VC, derives commitment, returns it
     *   5. Backend registers commitment on-chain (never seeing the DID)
     */
    func registerWithVC(vcJwt: String, countryCode: String) async throws -> String {
        let did   = try generateDID()
        let salt  = generateSalt()

        let body: [String: String] = [
            "vcJwt":       vcJwt,
            "countryCode": countryCode,
            "salt":        salt
        ]
        let response = try await apiClient.post(path: "/api/v1/did/commitment", body: body)
        guard let commitment = response["commitment"] as? String else {
            throw DIDError.invalidResponse
        }
        return commitment
    }

    // MARK: - Private

    private func generateSalt() -> String {
        var bytes = [UInt8](repeating: 0, count: 32)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes).hexString
    }

    /// Derive an Ethereum address from a P-256 public key.
    /// Note: Ethereum normally uses secp256k1; this is a compatible representation
    /// for did:ethr when using P-256 keys.
    private func ethereumAddress(from publicKey: P256.Signing.PublicKey) -> String {
        let keyBytes = publicKey.compressedRepresentation
        var hasher = SHA256()
        hasher.update(data: keyBytes)
        let hash = Data(hasher.finalize())
        // Take last 20 bytes as address (Ethereum convention)
        let addressBytes = hash.suffix(20)
        return "0x" + addressBytes.hexString
    }
}

enum DIDError: LocalizedError {
    case noSessionKey
    case invalidResponse
    case vcVerificationFailed

    var errorDescription: String? {
        switch self {
        case .noSessionKey:         return "No session key available"
        case .invalidResponse:      return "Invalid response from identity oracle"
        case .vcVerificationFailed: return "Verifiable credential verification failed"
        }
    }
}

private extension Data {
    var hexString: String { map { String(format: "%02x", $0) }.joined() }
}

File: android/app/src/main/kotlin/com/TheVoteApp/data/blockchain/DIDService.kt
package com.TheVoteApp.data.blockchain

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import com.TheVoteApp.data.remote.VoteApiClient
import java.math.BigInteger
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.MessageDigest
import java.security.SecureRandom
import java.security.interfaces.ECPublicKey
import javax.inject.Inject
import javax.inject.Singleton

/**
 * DIDService — W3C Decentralised Identity (did:ethr) client for Android.
 *
 * Generates a DID from the device's Android Keystore P-256 key and
 * derives the identity commitment for VoterRegistry registration.
 *
 * The private key never leaves the Android Keystore (hardware-backed on
 * supported devices). The DID is never transmitted to the backend —
 * only the commitment hash is sent.
 */
@Singleton
class DIDService @Inject constructor(
    private val apiClient: VoteApiClient,
    private val networkName: String = "sepolia"
) {

    companion object {
        private const val KEY_ALIAS = "TheVoteApp_did_key"
        private const val ANDROID_KEYSTORE = "AndroidKeyStore"
    }

    // MARK: - DID Generation

    /**
     * Generate a did:ethr DID from the device's Keystore public key.
     * The key is created once and reused across sessions.
     */
    @Throws(DIDException::class)
    fun generateDID(): String {
        ensureKeyExists()
        val address = deriveEthereumAddress()
        return "did:ethr:$networkName:$address"
    }

    // MARK: - Identity Commitment

    /**
     * Derive a voter identity commitment.
     * commitment = keccak256(did || countryCode || salt)
     *
     * Only the commitment is sent to the server — the DID stays on-device.
     */
    fun deriveIdentityCommitment(did: String, countryCode: String, salt: String): String {
        val input = (did + countryCode + salt).toByteArray(Charsets.UTF_8)
        val digest = MessageDigest.getInstance("SHA-256").digest(input)
        return digest.toHexString()
    }

    // MARK: - VC Registration

    /**
     * Register with the VoterRegistry using a government-issued
     * Verifiable Credential JWT.
     *
     * Flow:
     *   1. User receives a VC JWT from the national identity oracle
     *   2. App derives DID and salt locally
     *   3. App posts (vcJwt, countryCode, salt) to backend /did/commitment
     *   4. Backend verifies VC, derives commitment, returns it
     *   5. Backend registers commitment on-chain (never seeing the DID)
     */
    @Throws(DIDException::class)
    suspend fun registerWithVC(vcJwt: String, countryCode: String): String {
        val salt = generateSalt()
        val body = mapOf(
            "vcJwt" to vcJwt,
            "countryCode" to countryCode,
            "salt" to salt
        )
        val response = apiClient.post("/api/v1/did/commitment", body)
        return response["commitment"] as? String
            ?: throw DIDException.InvalidResponse
    }

    // MARK: - Private

    /**
     * Ensure the P-256 key pair exists in the Android Keystore.
     * Hardware-backed (StrongBox) when available; falls back to TEE.
     */
    private fun ensureKeyExists() {
        val keyStore = KeyStore.getInstance(ANDROID_KEYSTORE).apply { load(null) }
        if (keyStore.containsAlias(KEY_ALIAS)) return

        val spec = KeyGenParameterSpec.Builder(
            KEY_ALIAS,
            KeyProperties.PURPOSE_SIGN or KeyProperties.PURPOSE_VERIFY
        )
            .setAlgorithmParameterSpec(java.security.spec.ECGenParameterSpec("secp256r1"))
            .setDigests(KeyProperties.DIGEST_SHA256)
            .setUserAuthenticationRequired(false)
            .build()

        KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC, ANDROID_KEYSTORE)
            .apply { initialize(spec) }
            .generateKeyPair()
    }

    /**
     * Derive an Ethereum-style address from the Keystore public key.
     * Uses SHA-256 of the compressed public key bytes, taking the last 20 bytes.
     * Note: Ethereum normally uses secp256k1; this is a compatible representation
     * for did:ethr when using P-256 keys (same convention as iOS counterpart).
     */
    private fun deriveEthereumAddress(): String {
        val keyStore = KeyStore.getInstance(ANDROID_KEYSTORE).apply { load(null) }
        val publicKey = keyStore.getCertificate(KEY_ALIAS)?.publicKey as? ECPublicKey
            ?: throw DIDException.NoSessionKey

        val compressedKey = compressPublicKey(publicKey)
        val hash = MessageDigest.getInstance("SHA-256").digest(compressedKey)
        return "0x" + hash.takeLast(20).toByteArray().toHexString()
    }

    /** Compress a P-256 public key to 33 bytes (02/03 prefix + X coordinate). */
    private fun compressPublicKey(publicKey: ECPublicKey): ByteArray {
        val x = publicKey.w.affineX.toByteArrayUnsigned(32)
        val y = publicKey.w.affineY.toByteArrayUnsigned(32)
        val prefix = if (y.last().toInt() and 1 == 0) 0x02.toByte() else 0x03.toByte()
        return byteArrayOf(prefix) + x
    }

    private fun generateSalt(): String {
        val bytes = ByteArray(32)
        SecureRandom().nextBytes(bytes)
        return bytes.toHexString()
    }

    private fun ByteArray.toHexString(): String =
        joinToString("") { "%02x".format(it) }

    private fun List<Byte>.toByteArray(): ByteArray =
        ByteArray(size) { this[it] }

    private fun BigInteger.toByteArrayUnsigned(length: Int): ByteArray {
        val raw = toByteArray()
        return when {
            raw.size == length + 1 && raw[0] == 0.toByte() -> raw.drop(1).toByteArray()
            raw.size < length -> ByteArray(length - raw.size) + raw
            else -> raw
        }
    }
}

sealed class DIDException(message: String) : Exception(message) {
    object NoSessionKey : DIDException("No session key available in Android Keystore")
    object InvalidResponse : DIDException("Invalid response from identity oracle")
    object VCVerificationFailed : DIDException("Verifiable credential verification failed")
}
